#!/bin/Rscript
require(corrplot)
library(data.table)
library(dplyr)
library(qqman)
library(splines)

setwd("/home/richards/yiheng.chen/scratch/project1_2_metabolomics_GWAS_CLSA/codes/data")
CLSA_NORMDATAALL<-read.csv("./clsa_phenotypes/clsa_batch_norm_metabo_dat",na.strings=c("","NA"," ","Metabolite_not_called_in_this_set", "NaN"))

## metabolomics ADM_GWAS_ID
Metabolon_client_identifier<-read.csv("./clsa_phenotypes/linking_file")
## baseline data
Baseline_data_v3<-read.csv("./clsa_phenotypes/clsa_baseline_data", header=TRUE,stringsAsFactors=FALSE)
##metabolites annotation
matabolites_annotation<-read.csv("./clsa_phenotypes/clsa_metabo_annotation_dat")
#metabolomics metadata
ID_metadat<-read.csv("./clsa_phenotypes/clsa_metabo_meta_dat")
##subset baseline data
Baseline_data_v3_sub<-Baseline_data_v3 %>% select("ADM_GWAS3_COM","entity_id","SEX_ASK_COM", "AGE_NMBR_COM","BLD_ALC24_HR_COM","BLD_ALC24_MIN_COM", "BLD_FD24_HR_COM", "BLD_FD24_MIN_COM")
colnames(Baseline_data_v3_sub)<-c("ADM_GWAS_COM","entity_id","SEX_ASK_COM", "AGE_NMBR_COM", "BLD_ALC24_HR_COM","BLD_ALC24_MIN_COM", "BLD_FD24_HR_COM", "BLD_FD24_MIN_COM")
##### read in the ID for unrelated individuals (generated by King)
clsa_unrelated_indiv<-fread("./kingunrelated_list", header =F)

##remove the row without ADM_GWAS_COM
Baseline_data_v3_sub_withADMid<-Baseline_data_v3_sub[is.na(Baseline_data_v3_sub$ADM_GWAS_COM)==FALSE,]
##merge data
dat1<-merge(CLSA_NORMDATAALL, Metabolon_client_identifier, by="PARENT_SAMPLE_NAME")
dat2<-left_join(dat1, Baseline_data_v3_sub_withADMid, by="ADM_GWAS_COM")
dat3<-left_join(dat2, ID_metadat, by="PARENT_SAMPLE_NAME")

##### remove the metabolites have no measurements in >50% of samples
## get the list of metabolites with > 50% missing measurements
dat3_filtered_metaboMeasurement<-dat3 %>% select(X35:X999926111)
metabolite_list<-c()
for(i in names(dat3_filtered_metaboMeasurement)){
  if(sum(is.na(dat3_filtered_metaboMeasurement[[i]]))/length(dat3_filtered_metaboMeasurement[[i]])<0.5){
    metabolite_list<-append(metabolite_list,i)
  }
} ## 1091 metabolites pass this filter
dat4<-dat3 %>% select(PARENT_SAMPLE_NAME,metabolite_list,BLD_FD24_HR_COM,ADM_GWAS_COM)

#### check if there is any individual with over 50% missing
dat4_1<-dat4 %>% mutate(missingMeasurement_perce = rowSums(is.na(across(X35:X999926111)))/1091) ## all samples (9628 have all measurements)
dat4_filtered<-dat4 %>% mutate(missingMeasurement_perce = rowSums(is.na(across(X35:X999926111)))/1091) %>% filter(missingMeasurement_perce<0.5) ## all the individual have >50% measurements

#### get the individual with hour to last meal/drink information
dat5 <-dat4_filtered %>% filter(BLD_FD24_HR_COM>=0) ## 442 individuals have missing information were removed (left 9186)

nrow(dat5) ## 9186 samples have pass the filter 


################################# genomic information
fam_dat<-fread("./clsa_imputed_genotype_data/clsa_gen_v3.fam")
sample_dat<-fread("./clsa_imputed_genotype_data/clsa_imp_v3.sample")
sqc_file<-fread("./clsa_imputed_genotype_data/clsa_sqc_v3.txt")

table(sample_dat$ID_1 == sample_dat$ID_2)
# TRUE 
# 26631 

################ merge metabolomics data and genomic information
### list of individuals without metabolomics measurement or did not pass previous filter
ID_list<-data.frame(cbind(Metabolon_client_identifier$ADM_GWAS_COM, Metabolon_client_identifier$ADM_GWAS_COM))
                    ## get ID for individuals pass previous filter
ID_list_compl<-ID_list[complete.cases(ID_list),]
colnames(ID_list_compl)<-c("FID","IID")
ID_list_no_metabo<-sample_dat %>% select(ID_1, ID_2) %>% filter(!(ID_1 %in% ID_list_compl$FID) | !(ID_1 %in% dat5$ADM_GWAS_COM )) 
colnames(ID_list_no_metabo)<-c("FID","IID") ### 17445 individuals can be removed from GWAS

# setwd("/scratch/richards/yiheng.chen/project1_2_metabolomics_GWAS_CLSA/scratch/April122021_GCTA_test/April28_analysis_revised_clsa_metabo_data")
# write.table(ID_list_no_metabo,"indiv_no_metabo.txt",sep="\t",quote = FALSE, row.names = FALSE)

## prepare a new sample file
sample_dat_sub<-sample_dat %>% filter((ID_1 %in% ID_list_compl$FID) & (ID_1 %in% dat5$ADM_GWAS_COM ))

#### prepare covariates data for gwas (european only)
##get sqc data for european and unrelated individuals 
sqc_file_unrelated_europ<-sqc_file %>% 
  select(ADM_GWAS_COM, batch, chromosomal.sex,pca.cluster.id, ePC1, ePC2,ePC3, ePC4, ePC5, ePC6, ePC7, ePC8,ePC9, ePC10)%>% 
  filter((pca.cluster.id==4)& (ADM_GWAS_COM %in% sample_dat_sub$ID_1)) %>%
  filter(ADM_GWAS_COM %in% clsa_unrelated_indiv$V1)%>%
  mutate(Sex=ifelse(chromosomal.sex == 1, "M", ifelse(chromosomal.sex == 2, "F", "NA")))
#9186 (indiv with metabo measurement, food intake info) -->8502 (european)--> 8299 (no 2nd degree or closer relatives)


### check number of individuals from other ancestry
sqc_file_unrelated_all_ancestries<-sqc_file %>% 
  select(ADM_GWAS_COM, batch, chromosomal.sex,pca.cluster.id, PC1, PC2,PC3, PC4, PC5, PC6, PC7, PC8,PC9, PC10)%>% 
  filter(ADM_GWAS_COM %in% sample_dat_sub$ID_1) %>%
  filter(ADM_GWAS_COM %in% clsa_unrelated_indiv$V1)%>%
  mutate(Sex=ifelse(chromosomal.sex == 1, "M", ifelse(chromosomal.sex == 2, "F", "NA")))
table(sqc_file_unrelated_all_ancestries$pca.cluster.id)
#pca.cluster.id: 1 (South Asian)    2 (East Asian)    3 (Black)   4 (White)   5    6 
#n:              107                104               60          8299        341   68

### plot the PCA plots
sqc_file_unrelated_all_ancestries1<-sqc_file_unrelated_all_ancestries%>%rowwise()%>%mutate(ancestry = ifelse(pca.cluster.id ==1, "South Asian", ifelse(pca.cluster.id ==2, "East Asian", ifelse(pca.cluster.id ==3, "Black",ifelse(pca.cluster.id ==4, "White", ifelse(pca.cluster.id ==5, "White+ mixed","Latin American + mixed")) ))))
p<-ggplot(sqc_file_unrelated_all_ancestries1,aes(x=PC1,y=PC2,color=ancestry))
p<-p+geom_point(shape=1, size =3)+scale_color_manual(values=c('#e41a1c','#377eb8', '#4daf4a', "#984ea3", "#ff7f00", "#ffff33"))+
  theme_bw()+theme(axis.text.x = element_text(size=18), axis.text.y = element_text(size=18),
                   axis.title=element_text(size=20), 
                   legend.text=element_text(size=18), legend.title=element_text(size=20))
p
tiff("/scratch/richards/yiheng.chen/project1_2_metabolomics_GWAS_CLSA/scratch/June112021_GCTA_GWAS/July15201_revised_analysis/plotting_codes/00_PCA_plots_CLSA_individuals_with_metabolomics_measurements.tiff", width = 1000, height = 900, units = "px")
p
dev.off()

## prepare covariable files
covarCol_data<-cbind(sqc_file_unrelated_europ$ADM_GWAS_COM,sqc_file_unrelated_europ)
covarCol_data1<-left_join(covarCol_data, Baseline_data_v3_sub[,c(1,4)])
covarCol_data2<-left_join(covarCol_data1, dat5[,c(1093:1094)]) ## add BLD_FD24_HR_COM, ADM_GWAS_COM columns
colnames(covarCol_data2)<-c("FID","IID","batch","chromosomal.sex","pca.cluster.id","ePC1", "ePC2","ePC3","ePC4","ePC5","ePC6","ePC7","ePC8","ePC9","ePC10","Sex","Age", "BLD_FD24_HR_COM")

qcovarCol_clsa_europ_unrelated<-covarCol_data2 %>% select(FID, IID, ePC1, ePC2, ePC3, ePC4, ePC5, ePC6, ePC7, ePC8, ePC9, ePC10, Age, BLD_FD24_HR_COM) 
CovarCol_clsa_europ_unrelated <-covarCol_data2 %>% select(FID, IID, batch, Sex)
pheno
write.table(qcovarCol_clsa_europ_unrelated,"qcovarCol_clsa_europ_unrelated_noHeader.txt",sep="\t",quote = FALSE, row.names = FALSE, col.names = FALSE) ##8299 europeans have all information
write.table(CovarCol_clsa_europ_unrelated,"CovarCol_clsa_europ_unrelated_noHeader.txt",sep="\t",quote = FALSE, row.names = FALSE, col.names = FALSE)##8299 europeans have all information
write.table(CovarCol_clsa_europ_unrelated[,1:2],"/scratch/richards/yiheng.chen/project1_2_metabolomics_GWAS_CLSA/scratch/June112021_GCTA_GWAS/clsa_metabo_eur_unrelated_sublist_for_gwas_8299.txt",sep="\t",quote = FALSE, row.names = FALSE, col.names = FALSE)


#### prepare phenotype data for gwas
# fam_id<-data.frame(fam_dat[,1:2])
# colnames(fam_id)<-c("ADM_GWAS_COM","ADM_GWAS_COM1")

setwd("/home/richards/yiheng.chen/scratch/project1_2_metabolomics_GWAS_CLSA/codes/GWAS_related_results/metabolites_phenotype")
metabolite_list_totest<-unique(colnames(dat5[,2:1092]))
post_outlier_removal_mean_sd<-data.frame(
  metabolite=character(),
  mean=numeric(),
  sd=numeric(),
  standardized_mean=numeric(),
  standardized_sd=numeric())

for (metabo in metabolite_list_totest) {
    dat_sub<-dat5 %>% select(ADM_GWAS_COM,metabo)
    dat_sub$log_metabo<-apply(dat_sub, 1, function(x) log(as.numeric(x[2])))##natural log transform the data
    ##remove the measurements that 3 SD away
    mean_metabo<-mean(dat_sub$log_metabo,na.rm = TRUE)
    sd_metabo<-sd(dat_sub$log_metabo,na.rm = TRUE)
    dat_sub_filter<-dat_sub %>% filter(abs((log_metabo-mean_metabo)/sd_metabo)<=3)
    ##standardize data
    mean_metabo_post<-mean(dat_sub_filter$log_metabo, na.rm = TRUE)
    sd_metabo_post<-sd(dat_sub_filter$log_metabo, na.rm = TRUE)
    dat_sub_filter$std_metabo<-apply(dat_sub_filter, 1, function(x) (as.numeric(x[3])-mean_metabo_post)/sd_metabo_post)
    phenoCol<-left_join(covarCol_data[,1:2],dat_sub_filter[,c(1,4)],by="ADM_GWAS_COM")
    colnames(phenoCol)<-c("FID","IID","Standardized_metabolites_level")
    standardized_mean<-mean(phenoCol$Standardized_metabolites_level, na.rm = T)
    standardized_sd<-sd(phenoCol$Standardized_metabolites_level, na.rm = T)
    tmp_dat<-data.frame(metabo, mean_metabo_post, sd_metabo_post, standardized_mean, standardized_sd)
    colnames(tmp_dat)<-c("metabolite", "mean","sd", "standardized_mean", "standardized_sd")
    post_outlier_removal_mean_sd<-rbind(post_outlier_removal_mean_sd,tmp_dat)
    write.table(phenoCol, paste0("phenoCol_",metabo,"_clsa_europ_3sd.txt"),sep="\t",quote = FALSE,row.names = FALSE, col.names = FALSE)
}
